lib/db2.js

```diff
@@ -442,81 +443,94 @@ DB2.prototype.update = function(model, where, data, options, cb) {
-      for (var i = 0, n = fields.names.length; i < n; i++) {
-        setValues.push(new ParameterizedSQL('CAST (' + columnValues[i].sql +
-          ' AS ' +
-          self.buildColumnType(fields.properties[i]) + ')',
-          columnValues[i].params));
-
-        if (i < (n - 1))
-          setValues[i].sql = setValues[i].sql + ',';
-      }
-
-      sql.merge(setValues);
-      sql.merge('))) AS VT(' + fieldNames.join(',') + ')');
-      sql.merge('ON');
-      sql.merge('(MT.\"' + id + '\" = VT.\"' + id + '\")');
-      sql.merge('WHEN NOT MATCHED THEN INSERT (' +
-      fieldNames.join(',') + ')');
-
-      var values = ParameterizedSQL.join(columnValues, ',');
-      values.sql = 'VALUES(' + values.sql + ')';
-      sql.merge(values);
+    var stmt;
+    var tableName = self.tableEscaped(model);
+    var meta = {};
+
+    function executeWithConnection(connection, cb) {
+      // Execution for updateOrCreate requires running two
+      // separate SQL statements.  The second depends on the
+      // result of the first.
+      var countStmt = new ParameterizedSQL('SELECT COUNT(*) AS CNT FROM ');
+      countStmt.merge(tableName);
+      countStmt.merge(self.buildWhere(data));
+
+      connection.query(countStmt.sql, countStmt.params,
+        function(err, countData) {
+          if (err) {
+            return cb(err);
+          }
+
+          if (countData[0]['CNT'] > 0) {
+            stmt = self.buildUpdate(model, data.id, data);
+          } else {
+            stmt = self.buildInsert(model, data);
+          }
+
+          connection.query(stmt.sql, stmt.params, function(err, sData) {
+            if (err) {
+              return cb(err);
+            }
+
+            if (countData[0]['CNT'] > 0) {
+              meta.isNewInstance = false;
+              cb(null, data, meta);
+            } else {
+              stmt = 'SELECT MAX(' + id + ') as id FROM ' + tableName;
+              connection.query(stmt, null, function(err, info) {
+                if (err) {
+                  return cb(err);
+                }
+                data.id = info[0]['CNT'];
+                cb(null, data, meta);
+              });
+            }
+          });
+        });
+    };
+
+    if (options.transaction) {
+      executeWithConnection(options.transaction.connection,
+        function(err, data, meta) {
+          if (err) {
+            return callback && callback(err);
+          } else {
+            return callback && callback(null, data, meta);
+          }
+        });
     } else {
-      sql.merge(self.buildInsertDefaultValues(model, data, options));
-    }
-
-    sql.merge('WHEN MATCHED THEN UPDATE SET');
-
-    setValues = [];
-    for (i = 0, n = fields.names.length; i < n; i++) {
-      if (!fields.properties[i].id) {
-        setValues.push(new ParameterizedSQL(fields.names[i] + '=' +
-          '(' + columnValues[i].sql + ')', columnValues[i].params));
-      }
+      self.client.open(self.connStr, function(err, conn) {
```

Comment:
I wonder why you don't use https://github.com/strongloop/loopback-connector-db2/blob/zos/lib/transaction.js#L17. For example,
```
self.beginTransaction(err, conn) {
  ...
}
```